src/raft/recv_install_snapshot.c

```diff
@@ -8,6 +8,348 @@
 #include "recv.h"
 #include "replication.h"
 
+#include "../lib/sm.h"
+
+/**
+ * =Overview
+ *
+ * This detailed level design is based on PL018 and describes
+ * significant implementation details of data structures, RPCs
+ * introduced in it; provides model of operation and failure handling
+ * based on Leader's and Follower's states.
+ *
+ * =Data structures
+ *
+ * Among other structures it's needed to introduce a (persistent)
+ * container `HT` to efficiently store and map checksums to their page
+ * numbers on the follower's side. HT is implemented on top of sqlite3
+ * database with unix VFS. Every database corresponds to a
+ * raft-related database and maintains the following schema:
+ *
+ * CREATE TABLE "map" ("checksum" INTEGER NOT NULL, "pageno" INTEGER NOT NULL)
+ * CREATE INDEX map_idx on map(checksum);
+ *
+ * Each database stores a mapping from checksum to page number. This
+ * provides an efficient way to insert and lookup records
+ * corresponding to the checksums and page numbers.
+ */
+
+typedef unsigned int checksum_t;
+typedef unsigned long int pageno_t;
+
+struct page_checksum_t {
+	pageno_t   page_no;
+	checksum_t checksum;
+};
+
+struct raft_signature {
+	int version;
+
+	struct page_checksum_t *cs;
+	unsigned int cs_nr;
+	unsigned int cs_page_no;
+	const char *db;
+	int done;
+};
+
+struct raft_signature_result {
+	int version;
+	pageno_t last_known_page_no; /* used for retries and message losses */
+	int result;
+};
+
+struct page_from_to {
+	pageno_t from;
+	pageno_t to;
+};
+
+struct raft_install_snapshot_mv {
+	int version;
+
+	struct page_from_to *mv;
```

Comment:
Q: Why have a nested struct instead of inlining the fields? Asking out of pure curiosity.

Reply:
I meant having to separate fields in `raft_install_snapshot_mv` called `to` and `from`.